endpointmanager/pkg/chplquerier/productquerier.go

```diff
@@ -0,0 +1,313 @@
+	CriteriaMet         string `json:"criteriaMet"`
+	APIDocumentation    string `json:"apiDocumentation"`
+}
+
+// GetCHPLProducts queries CHPL for its HealthIT products using 'cli' and stores the products in 'store'
+// within the given context 'ctx'.
+func GetCHPLProducts(ctx context.Context, store endpointmanager.HealthITProductStore, cli *http.Client) error {
+	log.Debug("requesting products from CHPL")
+	prodJSON, err := getProductJSON(ctx, cli)
+	if err != nil {
+		return errors.Wrap(err, "getting health IT product JSON failed")
+	}
+	log.Debug("done requesting products from CHPL")
+
+	log.Debug("converting chpl json into product objects")
+	prodList, err := convertProductJSONToObj(ctx, prodJSON)
+	if err != nil {
+		return errors.Wrap(err, "converting health IT product JSON into a 'chplCertifiedProductList' object failed")
+	}
+	log.Debug("done converting chpl json into product objects")
+
+	log.Debug("persisting chpl products")
+	err = persistProducts(ctx, store, prodList)
+	log.Debug("done persisting chpl products")
+	return errors.Wrap(err, "persisting the list of retrieved health IT products failed")
+}
+
+// makes the request to CHPL and returns the byte string
+func getProductJSON(ctx context.Context, client *http.Client) ([]byte, error) {
+	chplURL, err := makeCHPLProductURL()
+	if err != nil {
+		return nil, errors.Wrap(err, "creating the URL to query CHPL failed")
+	}
+
+	// request ceritified products list
+	req, err := http.NewRequest("GET", chplURL.String(), nil)
+	if err != nil {
+		return nil, errors.Wrap(err, "creating http request failed")
+	}
+	req = req.WithContext(ctx)
+
+	resp, err := client.Do(req)
+	if err != nil {
+		return nil, errors.Wrap(err, "making the GET request to the CHPL server failed")
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return nil, errors.New("CHPL certified products request responded with status: " + resp.Status)
+	}
+
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, errors.Wrap(err, "reading the CHPL response body failed")
+	}
+
+	return body, nil
+}
+
+// creates the URL used to query CHPL including the data fields
+func makeCHPLProductURL() (*url.URL, error) {
+	queryArgs := make(map[string]string)
+	fieldStr := strings.Join(fields[:], ",")
+	queryArgs["fields"] = fieldStr
+
+	chplURL, err := makeCHPLURL(chplAPICertProdListPath, queryArgs)
+	if err != nil {
+		return nil, errors.Wrap(err, "creating the URL to query CHPL failed")
+	}
+
+	return chplURL, nil
+}
+
+// takes the json byte string and converts it into the associated JSON model
+func convertProductJSONToObj(ctx context.Context, prodJSON []byte) (*chplCertifiedProductList, error) {
+	var prodList chplCertifiedProductList
+
+	// don't unmarshal the JSON if the context has ended
+	select {
+	case <-ctx.Done():
+		return nil, errors.Wrap(ctx.Err(), "Unable to convert product JSON to objects - context ended")
+	default:
+		// ok
```

Comment:
Is there a reason why the default part of the select statement is kept here and elsewhere (like line 199)? It seems like you could just get rid of it, but I could see it being a placeholder for some handling you were considering for the future!

Reply:
the `default` is necessary. without it, the `select` statement will wait until a message shows up in one of its case statements.

Reply:
ah okay, i guess i misunderstood how the select statement worked, interesting!